### PR TITLE
Populate PackageInfo model with info from files.json

### PR DESCRIPTION
**Pull request by soapy1**

### CODE REVIEW

File: conda/models/package_info.py [hunk 1]
```diff
@@ -16,6 +17,10 @@
...
-
-    file = ListField(string_types)
-    # TODO: finish this
+    files = ListField(string_types)
+    has_prefix_files = ListField(string_types)
+    no_link = ListField(string_types)
+    soft_links = ListField(string_types)
+    index_json_record = ComposableField(Record)
+    icondata = StringField()
+    noarch = MapField()
```
Comment (kalefranz) [inline conda/models/package_info.py]: I don't like this format.  Thinking back to the `.psv` file, there should be one object for each row.  Let's discuss tomorrow.

**Comment by soapy1** [comment]
Note from Kale on https://github.com/soapy1/conda/pull/1

*Obviously this else part doesn't fit with what's above. We'll also have to modify the _make_link_operations code in conda/core/install.py that's currently using this.*

File: conda/models/package_info.py [hunk 2]
```diff
@@ -15,7 +19,30 @@
...
+    directory = 4
+
 
-    file = ListField(string_types)
-    # TODO: finish this
+class PathInfo(Entity):
+    path = StringField()
+    sha256 = StringField()
+    size_in_bytes = IntegerField()
+    file_type = EnumField(FileType)
```
Comment (soapy1) [inline conda/models/package_info.py]: @kalefranz  should this be https://github.com/conda/conda/blob/master/conda/base/constants.py#L63 instead?

File: conda/models/package_info.py [hunk 3]
```diff
@@ -2,9 +2,13 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (kalefranz) [inline conda/models/package_info.py]: We should have a `tests/models/test_package_info.py` file that is just a unit test or two that loads some contrived data.

File: conda/gateways/disk/read.py [hunk 4]
```diff
@@ -45,17 +46,44 @@ def collect_all_info_for_package(extracted_package_directory):
...
     files_path = join(extracted_package_directory, 'info', 'files')
-    files = tuple(ln for ln in (line.strip() for line in yield_lines(files_path)) if ln)
+    file_json_path = join(extracted_package_directory, 'info', 'files.json')
+
+    if isfile(file_json_path):
+        with open(file_json_path) as file_json:
+            data = json.load(file_json)
+        if data.get('version') != 1:
+            raise CondaUpgradeError("Expected files.json schema to be version 1")
+        return PackageInfo(files=(PathInfoV1(**f) for f in data['files']),
```
Comment (kalefranz) [inline conda/gateways/disk/read.py]: Just a small thing here... With all of these named arguments, I've found it better to create them one-per-line outside of the instance object creation.  That way, if/when you get a stack trace, you can more easily tell which of all of these function it's actually coming from.  So what I mean is something like

```
files=(PathInfoV1(**f) for f in data['files'])
index_json_record=read_index_json(extracted_package_directory)
noarch=read_noarch(extracted_package_directory)
icondata=read_icondata(extracted_package_directory)

return PackageInfo(files=files, index_json_record=index_json_record ...)
```

File: conda/exports.py [hunk 5]
```diff
@@ -80,7 +80,11 @@
 from conda.base.context import get_prefix as context_get_prefix, non_x86_linux_machines  # NOQA
 non_x86_linux_machines = non_x86_linux_machines
 
-from conda.base.constants import DEFAULT_CHANNELS  # NOQA
+from conda.base.constants import DEFAULT_CHANNELS       # NOQA
+from conda.models.package_info import NodeType          # NOQA
+from conda._vendor.auxlib.entity import EntityEncoder   # NOQA
```
Comment (kalefranz) [inline conda/exports.py]: We should be preferring relative imports here.  It's probably not done in the whole file.  No better time to start though.

File: conda/gateways/disk/read.py [hunk 6]
```diff
@@ -45,17 +46,44 @@ def collect_all_info_for_package(extracted_package_directory):
...
 
     files_path = join(extracted_package_directory, 'info', 'files')
-    files = tuple(ln for ln in (line.strip() for line in yield_lines(files_path)) if ln)
+    file_json_path = join(extracted_package_directory, 'info', 'files.json')
+
+    if isfile(file_json_path):
+        with open(file_json_path) as file_json:
+            data = json.load(file_json)
+        if data.get('version') != 1:
+            raise CondaUpgradeError("Expected files.json schema to be version 1")
```
Comment (kalefranz) [inline conda/gateways/disk/read.py]: "The current version of conda is too old to install this package. (This version only supports files.json schema version 1.)  Please update conda to install this package."

File: conda/core/install.py [hunk 7]
```diff
@@ -185,21 +188,24 @@ def _make_link_operations(self, requested_link_type):
         site_packages_dir = get_site_packages_dir(self.prefix)
         bin_dir = get_bin_directory_short_path()
 
-        def make_link_operation(source_short_path):
+        def make_link_operation(source_path_info):
             # no side effects in this method!
 
             # first part, same as parent class
-            if source_short_path in package_info.has_prefix_files:
+            if getattr(source_path_info, "prefix_placeholder", None) is not None:
```
Comment (kalefranz) [inline conda/core/install.py]: Here you could just add `required=False, nullable=True` to the `prefix_placeholder` field so you don't have to use `getattr`.

**Comment by kalefranz** [comment]
Any idea what's up with Travis-CI here?  I even restarted this build once.

Same problem here: https://github.com/conda/conda/pull/3887

**Comment by soapy1** [comment]
TravisCI is failing on Mac because the tests are taking to long to run - exceeding the max run time

**Comment by kalefranz** [comment]
@soap1 We definitely have broken tests here.  Not completely sure why.  I can look into it if I get to it first.

**Comment by kalefranz** [comment]
superseded by #3927

**Pull request closed without merging**